src/aerialmap_display.cpp

```diff
@@ -673,42 +868,100 @@ void AerialMapDisplay::transformTileToMapFrame()
+  center_tile_pose_.header.stamp = ref_fix_->header.stamp;
+  tf2::toMsg(t_navsat_map - t_centertile_navsat, center_tile_pose_.pose.position);
 }
 
-void AerialMapDisplay::transformMapTileToFixedFrame()
+void AerialMapDisplay::transformTileToUtmFrame()
 {
-  // orientation of the fixed-frame w.r.t. the map-frame
-  Ogre::Quaternion o_fixed_map;
-  // translation of the fixed-frame w.r.t. the map-frame
-  Ogre::Vector3 t_fixed_map;
+  if (!ref_fix_ or !center_tile_)
+  {
+    ROS_FATAL_THROTTLE_NAMED(2, "rviz_satellite", "ref_fix_ not set, can't create transforms");
+    return;
+  }
+  
+  // tile ID (integer x/y/zoom) corresponding to the downloaded tile / navsat message
+  auto const tile = fromWGSCoordinate<int>({ ref_fix_->latitude, ref_fix_->longitude }, zoom_);
+  
+  // Latitude and longitude of this tile's origin
+  auto const tileWGS = toWGSCoordinate<int>(tile, zoom_);
+
+  // Tile's origin in UTM coordinates
+  double northing, easting;
+  int utm_zone;
+  bool northp;
 
-  // get transform between map-frame and fixed-frame from the FrameManager
-  if (context_->getFrameManager()->getTransform(MAP_FRAME, ros::Time(), t_fixed_map, o_fixed_map))
+  try
   {
-    setStatus(::rviz::StatusProperty::Ok, "Transform", "Transform OK");
+    GeographicLib::UTMUPS::Forward(tileWGS.lat, tileWGS.lon, utm_zone, northp, easting, northing, utm_zone_);
+  }
+  catch (GeographicLib::GeographicErr& err)
+  {
+    ROS_ERROR_THROTTLE(2.0, "Error transforming lat-lon to UTM: %s", err.what());
+    if (utm_zone_ != GeographicLib::UTMUPS::STANDARD)
+    {
+      try
+      {
+        GeographicLib::UTMUPS::Forward(tileWGS.lat, tileWGS.lon, utm_zone, northp, easting, northing,
+                                       GeographicLib::UTMUPS::STANDARD);
+        ROS_INFO_THROTTLE(2.0, "Trying to autodetect UTM zone instead of using zone %i", utm_zone_);
+      }
+      catch (GeographicLib::GeographicErr& err)
+      {
+        setStatus(::rviz::StatusProperty::Error, "UTM", QString::fromStdString(err.what()));
+        return;
+      }
+    }
+  }
+  
+  setStatus(::rviz::StatusProperty::Ok, "UTM", "Conversion from lat/lon to UTM is OK.");
+
+  if (utm_zone != utm_zone_)
+  {
+    utm_zone_property_->setInt(utm_zone);
+  }
+
+  center_tile_pose_.header.stamp = ref_fix_->header.stamp;
+  center_tile_pose_.header.frame_id = utm_frame_;
+  center_tile_pose_.pose.position.x = easting;
+  center_tile_pose_.pose.position.y = northing;
+  center_tile_pose_.pose.position.z = ref_fix_->altitude;
+}
 
-    // the translation of the tile w.r.t. the fixed-frame
-    auto const t_centertile_fixed = t_fixed_map + o_fixed_map * t_centertile_map_;
+void AerialMapDisplay::transformMapTileToFixedFrame()
+{
+  // orientation of the tile w.r.t. the fixed-frame
+  Ogre::Quaternion o_centertile_fixed;
+  // translation of the tile w.r.t. the fixed-frame
+  Ogre::Vector3 t_centertile_fixed;
+
+  auto header = center_tile_pose_.header;
+  header.stamp = ros::Time::now();
```

Comment:
Sorry to bump up an old MR, but I'm curious about this change here. Is the use of `now()` necessary here as opposed to `ros::Time()` from before which should use the latest? I'm pretty sure this is causing issues for me when I try to use this plugin when the map frame and fixed frame are not the same (due to tf lookup errors).

Reply:
Are you using the "UTM frame" mode or the default "Map frame" mode?

Reply:
I'm using Map Frame mode.

Reply:
Thanks. And what kind of tf lookup errors do you get. Can you post their text here, as well as your RViz configuration (fixed frame + AerialMapDisplay config)?

Reply:
So the original error was cannot find transform due to an unknown reason. But this turned out to be a red herring because the call to hasProblems uses a ros::Time() but the call to transform uses ros::Time::now(). So hasProblems would return false and the unknown error message would get displayed. After I passed header.stamp into hasProblems, the error message in rviz was "getting transform would require extrapolating into the future", by some small amount. The takeaway is that passing ros::Time() (zero time) asks for the most recent available transform, but using now() can fail. So, extrapolating / poorly synchronized messages aside, I think this should continue to use ros::Time(), to avoid complete failure in that case. Was ros Time now used to get better alignment?

Reply:
Hmm, the affected method is called inside `update()` function (the RViz UI redraw callback), so it definitely cannot wait for the transform to become available. So I'm inclined to agree with you that using `ros::Time()` is correct here.

I'll try to make that change and run in UTM mode (as I haven't been able to use Map Frame Mode sanely) and check if I see something weird. Can you test locally with this change in Map Frame Mode?

Just to get a bit more context, what is your map frame and fixed frame? How are they created?

@Timple any thoughts?